### PR TITLE
HPCC-29283 Alternative incr.sh

### PR DESCRIPTION
**Pull request by GordonSmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by GordonSmith** [comment]
@richardkchapman @jakesmith @mckellyln This should be the finalised version of the new incr.sh replacement.

Could you take a moment to test using:  `./dockerfiles/image.sh incr`, then:
1.  Make a change and re-run
2.  Just re-run (should be cached)
3. Make a "release" image:  `./dockerfiles/image.sh build`
4. Make a "debug" image:  `./dockerfiles/image.sh build -m debug`

@mckellyln - this script also supports targeting diff OS for image (like Ubuntu 20.04), but I will need to add a platform-core-ubuntu-20.04 dockerfile (probably the same or very similar to the 22.04 variant).

**Review by richardkchapman — APPROVED** [review]
Looks ok other than one potentially unrelated change included

File: .gitignore [hunk 1]
```diff
@@ -17,4 +17,5 @@ eclcc.log
 ./helm/examples/azure/sa/env-sa
 ./dockerfiles/platform-build-incremental/hpcc.gitpatch
 .env
-vcpkg.json
\ No newline at end of file
+vcpkg.json
+rsync_include.txt
```
Comment (GordonSmith) [inline .gitignore:21]: This could be moved into one of the mounted volumes so wouldn't need this exclusion.

File: dockerfiles/image.sh [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+build() {
+    if [ "$MODE" = "release" ]; then
+        local label=platform-core
+        local base=ubuntu:jammy-20230308
+        local build_type="RelWithDebInfo"
+        local cmake_options="-DCMAKE_BUILD_TYPE=$build_type -DCPACK_STRIP_FILES=ON"
+    elif [ "$MODE" = "debug" ]; then
+        local label=platform-core-debug
```
Comment (jakesmith) [inline dockerfiles/image.sh:159]: minor: it's going to be slightly annoying that the debug version has a different image name.
In the past the tag would be different, but incr.sh would always build "platform-core".

Being different means startall.sh will not automatically pick up the latest "platform-core-debug" (it's only looking for platform-core).

**Review by jakesmith — COMMENTED** [review]
Looks nice, a few comments to consider.

File: dockerfiles/image.sh [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        echo "  --- Incremental '$image_name' image ---"
+    fi
+    CONTAINER=$(docker run -d \
+        --mount source=hpcc_src,target=/hpcc-dev/HPCC-Platform,type=volume \
+        --mount source=$HPCC_BUILD,target=/hpcc-dev/build,type=volume \
+        $image_name "tail -f /dev/null")
+    docker exec --user root $CONTAINER /bin/bash -c "$cmd"
+    docker exec --user root $CONTAINER /bin/bash -c "eclcc -pch"
+    docker commit $CONTAINER hpccsystems/$label:$GITHUB_BRANCH-$crc
+    docker commit $CONTAINER hpccsystems/$label:latest
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply dockerfiles/image.sh]: Yea - I had already made that change!

File: dockerfiles/image.sh [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        echo "hpccsystems/$label:$GITHUB_BRANCH-$crc"
+        exit 0
+    fi
+}
+
+build() {
+    if [ "$MODE" = "release" ]; then
+        local label=platform-core
+        local base=ubuntu:jammy-20230308
+        local build_type="RelWithDebInfo"
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply dockerfiles/image.sh:159]: Thanks - just to clarify, for bare metal its RelWithDebInfo, but for k8s image it should be Release?

File: dockerfiles/image.sh [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    fi
+
+    if [ "$MODE" = "release" ]; then
+        run "cmake --build /hpcc-dev/build --parallel --target package"
+        finalize_platform_core_image $label $crc \
+            "dpkg -i /hpcc-dev/build/hpccsystems-platform*.deb"
+    elif [ "$MODE" = "debug" ]; then
+        run "cmake --build /hpcc-dev/build --parallel"
+        finalize_platform_core_image $label $crc \
+            "cmake --build /hpcc-dev/build --parallel --target install"
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply dockerfiles/image.sh:202]: 1. Building the package is slow and install has the added benefit of only installing changed files.
2. In release mode, the platform-core image does not have any cmake or ninja tools so can't do an "install".  Also I plan on adding the facility to install an extern deb file (for GH build-assets action)

File: dockerfiles/image.sh [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    echo "Usage: $0 [-h] {clean|build|incr} [-m MODE] [-r] [OS]"
+    echo "  -h, --help     display this help message"
+    echo "  clean          remove all build artifacts"
+    echo "  build          build the project"
+    echo "  incr           perform an incremental build (faster version of 'build -m debug')"
+    echo "  status         display environment variables"
+    echo "  -m, --mode     specify the build mode (debug or release)"
+    echo "                 default mode is release"
+    echo "  -r, --reconfigure reconfigure CMake before building"
+}
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply dockerfiles/image.sh:235]: Will do as separate issue/pr

File: dockerfiles/image.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply dockerfiles/image.sh]: I think it's the other way around atm:

Default is "Release" in both: https://github.com/hpcc-systems/HPCC-Platform/blob/b6c727364dd4844e13e13cdfe9be4ac65ac1790a/cmake_modules/options.cmake#L97

but in k8s and dockerfiles/platform-build, it overrides the default to "RelWithDebInfo": https://github.com/hpcc-systems/HPCC-Platform/blob/b6c727364dd4844e13e13cdfe9be4ac65ac1790a/dockerfiles/platform-build/Dockerfile#L84
(I don't see where the GH actions override that).

But if anything they should be other way around, "Release" in image for size reasons.

**Comment by jakesmith** [comment]
> https://github.com/hpcc-systems/HPCC-Platform/pull/17176#discussion_r1158351505

I see the change from GIT_REF to GIT_BRANCH, but isn't this image invariant on branch ?
Could it instead be tagged and reference via $VCPKG_REF?

**Review by jakesmith — APPROVED** [review]
@GordonSmith - looks good.

A couple of queries, but wouldn't hold up merging it.

**Comment by GordonSmith** [comment]
> 



> > [#17176 (comment)](https://github.com/hpcc-systems/HPCC-Platform/pull/17176#discussion_r1158351505)
> 
> I see the change from GIT_REF to GIT_BRANCH, but isn't this image invariant on branch ? Could it instead be tagged and reference via $VCPKG_REF?

It could be and it does a have a direct dependency on that version...  I will make that change.

**Review by jakesmith — APPROVED** [review]
@GordonSmith - changes look good.

**Comment by GordonSmith** [comment]
Rebased due to conflict.

@Michael-Gardner FYI

**Review by mckellyln — APPROVED** [review]
Looks very good to me, I need to use it to maybe see about any questions.
I say ok to merge in.

**Pull request merged**